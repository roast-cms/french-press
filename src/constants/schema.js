import {Block} from "slate"

/**
 * A set of rules that transform document structure and keep it normalized to a defined format.
 * @constant schema
 * @property {Array} nodes Acceptable nodes within editor document.
 * @property {Object} last Defines at least one empty paragraph block that follows a void block (such as picture); this is required to ensure that the user can continue adding content without additional effort below uploaded images (otherwise they will be forced to move the image up to free up a trailing paragraph space).
 */

export const SCHEMA = {
  document: {
    nodes: [
      {
        types: [
          "paragraph",
          "heading",
          "divider",
          "quote",
          "image",
          "docket",
          "link",
        ],
      },
    ],
    last: {types: ["paragraph", "quote"]},
    normalize: (change, reason, {node}) => {
      switch (reason) {
        case "last_child_type_invalid": {
          const paragraph = Block.create("paragraph")
          return change.insertNodeByKey(node.key, node.nodes.size, paragraph)
        }
        default:
          return null
      }

    },
  },
  blocks: {
    link: {
      nodes: [{objects: ["text"]}],
    },
    divider: {
      isVoid: true,
    },
    image: {
      isVoid: true,
      data: {
        src: value => value,
      },
      parent: [{object: "document"}],
      normalize: (change, {code, node}) => {
        if (code === "parent_object_invalid") {
          return change.unwrapBlockByKey(node.key)
        } else return null
      },
    },
    docket: {
      isVoid: true,
    },
  },
  inlines: {
    quote: {
      nodes: [{types: ["text"]}],
    },
    paragraph: {
      nodes: [{types: ["text", "link"]}],
    },
    heading: {
      nodes: [{types: ["text"]}],
    },
  },
}
