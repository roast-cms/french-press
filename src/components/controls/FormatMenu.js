import React from "react"
import styled from "styled-components"

import button from "./css/button"

const Menu = styled.div`
  display: none;
  position: absolute;
  bottom: initial !important;
  z-index: ${props => props.theme.layer.up + 1};
  width: auto;

  background: #7b294a;
  color: #fff;
  border-radius: 0.25em;
  overflow: hidden;

  button {
    ${button}
  }

  &.touch {
    margin-top: -65px;
    ${props => props.theme.size.breakpoint.min.m`
    &::after {
      content: "";
      position: absolute;
      display: block;
      height: 57px;
      margin-left: calc(5em - 1px);
      margin-top: 2px;
      border: 1px dashed ${props => props.theme.color.brand};
      z-index: -1;
    }`};
  }
  ${props => props.theme.size.breakpoint.max.s`
    width: 100vw;
    left: 0 !important;
    &.touch {
      margin-top: -55px;
    }
    & > div {
      background: ${props =>
        props.theme.color.foreground(props.theme.opacity.most)};
    }
  `};
`
const MenuItem = styled.button`
  width: auto;
  flex-grow: 1;
`

/**
 * A component with all of the UI logic for the floating format menu.
 * @module FormatMenu
 */
export default props => {
  if (props.value.blocks.some(node => node.type === "quote")) return null
  if (
    props.value.blocks.some(node => node.type === "heading") &&
    props.value.blocks.some(
      node => node.type === "paragraph" && node.text !== ""
    )
  )
    return null
  const MakeHeader =
    props.controls && props.controls.MakeHeader
      ? props.controls.MakeHeader
      : () => <span>H</span>
  const CancelHeader =
    props.controls && props.controls.CancelHeader
      ? props.controls.CancelHeader
      : () => <small>⇲</small>
  const MakeQuote =
    props.controls && props.controls.MakeQuote
      ? props.controls.MakeQuote
      : () => <span>❝</span>
  const MakeLink =
    props.controls && props.controls.MakeLink
      ? props.controls.MakeLink
      : () => <u>a</u>
  const MakeBold =
    props.controls && props.controls.MakeBold
      ? props.controls.MakeBold
      : () => <strong>b</strong>
  const MakeItalic =
    props.controls && props.controls.MakeItalic
      ? props.controls.MakeItalic
      : () => <i>i</i>
  //
  return (
    <Menu innerRef={props.menuRef}>
      {props.value.blocks.some(node => node.type === "heading") ? (
        <div
          style={{display: "block"}}
          className="french-press_undo-heading-container"
        >
          <button
            onMouseDown={event => event.preventDefault()}
            onMouseUp={event => {
              event.preventDefault()
              props.formatCommand("undo_heading")
            }}
            branded
            style={{
              width: "1.55em",
            }}
            title="Undo heading"
            className="french-press_undo-heading"
          >
            <CancelHeader />
          </button>
        </div>
      ) : (
        <div style={{display: "flex"}}>
          <MenuItem
            branded
            left
            title="Make a heading"
            onMouseDown={event => event.preventDefault()}
            onMouseUp={event => {
              event.preventDefault()
              props.formatCommand("make_heading")
            }}
          >
            <MakeHeader />
          </MenuItem>
          <MenuItem
            branded
            title="Make a quote"
            onMouseDown={event => event.preventDefault()}
            onMouseUp={event => {
              event.preventDefault()
              props.formatCommand("make_quote")
            }}
          >
            <MakeQuote />
          </MenuItem>
          <MenuItem
            script
            branded={!props.value.inlines.some(node => node.type === "link")}
            inverse={props.value.inlines.some(node => node.type === "link")}
            onMouseDown={event => event.preventDefault()}
            onMouseUp={event => {
              event.preventDefault()
              props.formatCommand("toggle_link")
            }}
            style={{borderLeft: "4px solid #2c2c2c"}}
          >
            <MakeLink />
          </MenuItem>
          <MenuItem
            script
            branded={
              props.value &&
              props.value.activeMarks &&
              !props.value.activeMarks.some(mark => mark.type === "bold")
            }
            inverse={
              props.value &&
              props.value.activeMarks &&
              props.value.activeMarks.some(mark => mark.type === "bold")
            }
            onClick={event => event.preventDefault()}
            onMouseDown={event => event.preventDefault()}
            onMouseUp={event => {
              event.preventDefault()
              props.formatCommand("toggle_bold")
            }}
          >
            <MakeBold style={{fontWeight: "700 !important"}} />
          </MenuItem>
          <MenuItem
            script
            branded={
              props.value &&
              props.value.activeMarks &&
              !props.value.activeMarks.some(mark => mark.type === "italic")
            }
            inverse={
              props.value &&
              props.value.activeMarks &&
              props.value.activeMarks.some(mark => mark.type === "italic")
            }
            right
            onClick={event => event.preventDefault()}
            onMouseDown={event => event.preventDefault()}
            onMouseUp={event => {
              event.preventDefault()
              props.formatCommand("toggle_italic")
            }}
          >
            <MakeItalic />
          </MenuItem>
        </div>
      )}
    </Menu>
  )
}
