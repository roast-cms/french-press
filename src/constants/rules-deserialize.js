/**
 * A set of rules that defines how to transpile HTML itno slate document nodes and marks (and back).
 * @constant rules
 */
import isUrl from "is-url"

import {BLOCK_TAGS, MARK_TAGS} from "./defaults"

/**
 * Flattens HTML into plain text.
 * @function squish
 * @return {String}
 */
const squish = el => {
  el.innerHTML = el.innerText || el.textContent
  return el
}


export const RULES_DESERIALIZE = [
  {
    deserialize(el, next) {
      const block = BLOCK_TAGS[el.tagName.toLowerCase()]
      if (!block) return
      switch (block) {
        case "paragraph": {
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes),
          }
        }
        case "quote": {
          return {
            object: "block",
            type: "quote",
            nodes: next(el.childNodes),
          }
        }
        case "heading": {
          return {
            object: "block",
            type: "heading",
            nodes: next(el.childNodes),
          }
        }
        case "image": {
          const src = el.getAttribute("src") || el.getAttribute("srcset")
          if (!isUrl(src)) return
          return {
            object: "block",
            type: "image",
            isVoid: true,
            data: {src},
          }
        }
        case "link": {
          return {
            object: "inline",
            type: "link",
            data: {
              href: el.getAttribute("href"),
            },
            nodes: next(el.childNodes),
          }
        }
        default:
          return {
            object: "block",
            type: "paragraph",
            nodes: next(el.childNodes),
          }
      }
    },
  },
  {
    deserialize(el, next) {
      const mark = MARK_TAGS[el.tagName]
      if (!mark) return
      return {
        object: "mark",
        type: mark,
        nodes: next(el.childNodes),
      }
    },
  },
]
