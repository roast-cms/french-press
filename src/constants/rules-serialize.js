import React from "react"

import {addKey} from "../components/vignettes/Reader"
import Link from "../components/controls/Link"

export const rulesSerializeWithProps = props => [
  {
    serialize(node, children) {
      const element = node.type
      switch (element) {
        case "link": {
          const href = node.data.href
          return addKey(
            <Link
              to={href}
              domain={props.options ? props.options.domain : null}
            >
              {children}
            </Link>
          )
        }
        case "image": {
          if (props.components && props.components.Picture) {
            const Picture = props.editor
              ? props.editor.props.components.Picture
              : props.components.Picture
            return addKey(
              <Picture
                editor={{value: {isSelected: false}}}
                node={{
                  data: {
                    get: object => node.data[object],
                  },
                  serial: node.serial,
                }}
                readOnly={true}
              />
            )
          } else return null
        }
        default:
          return addKey(<span>{children}</span>)
      }
    },
  },
]
export const RULES_SERIALIZE = [
  {
    serialize(node, children) {
      const element = node.type
      switch (element) {
        case "paragraph": {
          return addKey(<p>{children}</p>)
        }
        case "quote": {
          return addKey(
            <div style={{clear: "both"}}>
              <blockquote>{children}</blockquote>
            </div>
          )
        }
        case "heading": {
          return addKey(<h3>{children}</h3>)
        }
        case "divider": {
          return addKey(<hr />)
        }
        case "italic": {
          return addKey(<em>{children}</em>)
        }
        case "bold": {
          return addKey(<strong>{children}</strong>)
        }
      }
    },
  },
]
