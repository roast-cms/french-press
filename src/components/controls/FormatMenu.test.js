import "jest-styled-components"

import {Value} from "slate"
import React from "react"

import {shallow} from "enzyme"

import {DEFAULT_EDITOR_STATE} from "../../constants"
import FormatMenu from "./FormatMenu"

test("Render FormatMenu without crashing, matches snapshot", () => {
  const element = shallow(
    <FormatMenu value={Value.fromJSON(DEFAULT_EDITOR_STATE)} />
  )
  expect(element).toMatchSnapshot()
})
