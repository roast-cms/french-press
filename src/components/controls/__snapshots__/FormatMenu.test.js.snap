// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <_default
    value={
      Immutable.Record {
        "data": Immutable.Map {},
        "decorations": null,
        "document": Immutable.Record {
          "data": Immutable.Map {},
          "key": "1",
          "nodes": Immutable.List [
            Immutable.Record {
              "data": Immutable.Map {},
              "isVoid": false,
              "key": "2",
              "nodes": Immutable.List [
                Immutable.Record {
                  "leaves": Immutable.List [
                    Immutable.Record {
                      "marks": Immutable.Set [],
                      "text": "",
                    },
                  ],
                  "key": "3",
                },
              ],
              "type": "paragraph",
            },
          ],
        },
        "history": Immutable.Record {
          "redos": Immutable.Stack [],
          "undos": Immutable.Stack [],
        },
        "schema": Immutable.Record {
          "stack": Immutable.Record {
            "plugins": Array [
              Object {
                "validateNode": [Function],
              },
              Object {
                "validateNode": [Function],
              },
              Object {
                "validateNode": [Function],
              },
              Object {
                "validateNode": [Function],
              },
              Object {
                "validateNode": [Function],
              },
              Object {
                "validateNode": [Function],
              },
              Object {
                "validateNode": [Function],
              },
              Object {
                "validateNode": [Function],
              },
              Object {
                "schema": Object {},
              },
            ],
          },
          "document": Object {
            "data": Object {},
            "nodes": null,
          },
          "blocks": Object {},
          "inlines": Object {},
        },
        "selection": Immutable.Record {
          "anchorKey": "3",
          "anchorOffset": 0,
          "anchorPath": Immutable.List [
            0,
            0,
          ],
          "focusKey": "3",
          "focusOffset": 0,
          "focusPath": Immutable.List [
            0,
            0,
          ],
          "isAtomic": false,
          "isBackward": false,
          "isFocused": false,
          "marks": null,
        },
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <styled.div>
        <div
          style={
            Object {
              "display": "flex",
            }
          }
        >
          <styled.button
            left={true}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            title="Make a heading"
          >
            <Unknown />
          </styled.button>
          <styled.button
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            title="Make a quote"
          >
            <Unknown />
          </styled.button>
          <styled.button
            className={null}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            script={true}
            style={
              Object {
                "borderLeft": "4px solid #2c2c2c",
              }
            }
          >
            <Unknown />
          </styled.button>
          <styled.button
            className={null}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            script={true}
          >
            <Unknown
              style={
                Object {
                  "fontWeight": "700 !important",
                }
              }
            />
          </styled.button>
          <styled.button
            className={null}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            right={true}
            script={true}
          >
            <Unknown />
          </styled.button>
        </div>
      </styled.div>,
      "className": "fpe-menu",
      "innerRef": undefined,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <div
          style={
            Object {
              "display": "flex",
            }
          }
        >
          <styled.button
            left={true}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            title="Make a heading"
          >
            <Unknown />
          </styled.button>
          <styled.button
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            title="Make a quote"
          >
            <Unknown />
          </styled.button>
          <styled.button
            className={null}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            script={true}
            style={
              Object {
                "borderLeft": "4px solid #2c2c2c",
              }
            }
          >
            <Unknown />
          </styled.button>
          <styled.button
            className={null}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            script={true}
          >
            <Unknown
              style={
                Object {
                  "fontWeight": "700 !important",
                }
              }
            />
          </styled.button>
          <styled.button
            className={null}
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseUp={[Function]}
            right={true}
            script={true}
          >
            <Unknown />
          </styled.button>
        </div>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <styled.button
              left={true}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              title="Make a heading"
            >
              <Unknown />
            </styled.button>,
            <styled.button
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              title="Make a quote"
            >
              <Unknown />
            </styled.button>,
            <styled.button
              className={null}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              script={true}
              style={
                Object {
                  "borderLeft": "4px solid #2c2c2c",
                }
              }
            >
              <Unknown />
            </styled.button>,
            <styled.button
              className={null}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              script={true}
            >
              <Unknown
                style={
                  Object {
                    "fontWeight": "700 !important",
                  }
                }
              />
            </styled.button>,
            <styled.button
              className={null}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              right={true}
              script={true}
            >
              <Unknown />
            </styled.button>,
          ],
          "style": Object {
            "display": "flex",
          },
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Unknown />,
              "left": true,
              "onMouseDown": [Function],
              "onMouseUp": [Function],
              "title": "Make a heading",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Unknown />,
              "onMouseDown": [Function],
              "onMouseUp": [Function],
              "title": "Make a quote",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Unknown />,
              "className": null,
              "onMouseDown": [Function],
              "onMouseUp": [Function],
              "script": true,
              "style": Object {
                "borderLeft": "4px solid #2c2c2c",
              },
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Unknown
                style={
                  Object {
                    "fontWeight": "700 !important",
                  }
                }
              />,
              "className": null,
              "onClick": [Function],
              "onMouseDown": [Function],
              "onMouseUp": [Function],
              "script": true,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "style": Object {
                  "fontWeight": "700 !important",
                },
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "children": <Unknown />,
              "className": null,
              "onClick": [Function],
              "onMouseDown": [Function],
              "onMouseUp": [Function],
              "right": true,
              "script": true,
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            "type": [Function],
          },
        ],
        "type": "div",
      },
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <styled.div>
          <div
            style={
              Object {
                "display": "flex",
              }
            }
          >
            <styled.button
              left={true}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              title="Make a heading"
            >
              <Unknown />
            </styled.button>
            <styled.button
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              title="Make a quote"
            >
              <Unknown />
            </styled.button>
            <styled.button
              className={null}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              script={true}
              style={
                Object {
                  "borderLeft": "4px solid #2c2c2c",
                }
              }
            >
              <Unknown />
            </styled.button>
            <styled.button
              className={null}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              script={true}
            >
              <Unknown
                style={
                  Object {
                    "fontWeight": "700 !important",
                  }
                }
              />
            </styled.button>
            <styled.button
              className={null}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              right={true}
              script={true}
            >
              <Unknown />
            </styled.button>
          </div>
        </styled.div>,
        "className": "fpe-menu",
        "innerRef": undefined,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <div
            style={
              Object {
                "display": "flex",
              }
            }
          >
            <styled.button
              left={true}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              title="Make a heading"
            >
              <Unknown />
            </styled.button>
            <styled.button
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              title="Make a quote"
            >
              <Unknown />
            </styled.button>
            <styled.button
              className={null}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              script={true}
              style={
                Object {
                  "borderLeft": "4px solid #2c2c2c",
                }
              }
            >
              <Unknown />
            </styled.button>
            <styled.button
              className={null}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              script={true}
            >
              <Unknown
                style={
                  Object {
                    "fontWeight": "700 !important",
                  }
                }
              />
            </styled.button>
            <styled.button
              className={null}
              onClick={[Function]}
              onMouseDown={[Function]}
              onMouseUp={[Function]}
              right={true}
              script={true}
            >
              <Unknown />
            </styled.button>
          </div>,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <styled.button
                left={true}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title="Make a heading"
              >
                <Unknown />
              </styled.button>,
              <styled.button
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                title="Make a quote"
              >
                <Unknown />
              </styled.button>,
              <styled.button
                className={null}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                script={true}
                style={
                  Object {
                    "borderLeft": "4px solid #2c2c2c",
                  }
                }
              >
                <Unknown />
              </styled.button>,
              <styled.button
                className={null}
                onClick={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                script={true}
              >
                <Unknown
                  style={
                    Object {
                      "fontWeight": "700 !important",
                    }
                  }
                />
              </styled.button>,
              <styled.button
                className={null}
                onClick={[Function]}
                onMouseDown={[Function]}
                onMouseUp={[Function]}
                right={true}
                script={true}
              >
                <Unknown />
              </styled.button>,
            ],
            "style": Object {
              "display": "flex",
            },
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Unknown />,
                "left": true,
                "onMouseDown": [Function],
                "onMouseUp": [Function],
                "title": "Make a heading",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Unknown />,
                "onMouseDown": [Function],
                "onMouseUp": [Function],
                "title": "Make a quote",
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Unknown />,
                "className": null,
                "onMouseDown": [Function],
                "onMouseUp": [Function],
                "script": true,
                "style": Object {
                  "borderLeft": "4px solid #2c2c2c",
                },
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Unknown
                  style={
                    Object {
                      "fontWeight": "700 !important",
                    }
                  }
                />,
                "className": null,
                "onClick": [Function],
                "onMouseDown": [Function],
                "onMouseUp": [Function],
                "script": true,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "style": Object {
                    "fontWeight": "700 !important",
                  },
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "children": <Unknown />,
                "className": null,
                "onClick": [Function],
                "onMouseDown": [Function],
                "onMouseUp": [Function],
                "right": true,
                "script": true,
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {},
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              "type": [Function],
            },
          ],
          "type": "div",
        },
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
